Blog and docs on new TF converter (#3011)
pulumi/pulumi-hugo: themes/default/content/blog/converting-full-terraform-programs-to-pulumi/index.md

```

{{% /choosable %}}
{{% choosable language python %}}

```python
alb = aws.lb.LoadBalancer(f"{name}-alb",
    name=pulumi.Output.from_input(args["namePrefix"]).apply(lambda np: np.replace("_", "-")),
```

{{% /choosable %}}
{{% choosable language go %}}

```go
alb, err := lb.NewLoadBalancer(ctx, fmt.Sprintf("%s-alb", name), &lb.LoadBalancerArgs{
    Name: args.NamePrefix.ToStringOutput().ApplyT(func(name string) string {
        return strings.ReplaceAll(name, "_", "-")
    }).(pulumi.StringOutput),
```

{{% /choosable %}}
{{% choosable language csharp %}}

```csharp
var alb = new Aws.Lb.LoadBalancer($"{name}-alb", new()
{
    Name = args.NamePrefix.Apply(name => name.Replace("_", "-"));
```

{{% /choosable %}}
{{< /chooser >}}

There are some other TODOs for missing `merge` and `template_file` support, which also need to be filled in. After addressing these and some other tweaks to make the code compile, we can now run the Pulumi program with `pulumi up` to provision the Vault installation with Pulumi.

The converter has saved us a ton of time compared to doing the migration entirely by hand!

## Wrapping Up

> The new conversion support makes it easy to modernize your legacy IaC to take advantage of the benefits of Pulumi's deployment engine, programming model, and developer productivity! We are actively working on improving the new Terraform converter -- reducing the number of emitted TODOs and improving the overall quality of the generated code. If you have existing Terraform code that you need to migrate, give the new converter a try, and [let us know](https://github.com/pulumi/pulumi/issues/new/choose) if you run into any issues.

Reviewer: suggest ending with a CTA for getting started with it

---
title: "Converting Full Terraform Programs to Pulumi"
date: 2023-06-12
meta_desc: Learn how to convert whole Terraform programs to Pulumi using the new Terraform converter

meta_image: meta.png

authors:
    - justin-vanpatten

tags:
    - migration
---

Today, we're excited to announce support for converting whole Terraform projects to Pulumi via the `pulumi convert` command in the Pulumi CLI. The new Terraform converter includes support for Terraform modules, all core features of Terraform 1.4, and the majority of Terraform built-in functions. Let's dig in to learn more about the new converter and how to use it.

<!--more-->

> Historically, we have offered a separate [tf2pulumi](https://github.com/pulumi/tf2pulumi) tool to convert snippets of Terraform to Pulumi. Today we're announcing a new converter that supports converting the vast majority of complete Terraform projects, including modules. There's no longer a separate tool to download -- you can run the new converter directly from the Pulumi CLI with the `pulumi convert --from terraform` command to convert Terraform to Pulumi TypeScript, Python, Go, or C#.

Reviewer 2: > There's no longer a separate tool to download [...] you can run the new converter

I suggest adding `As of v{current version}, you can run the new converter directly from the Pulumi CLI`

Reviewer 3: Yeh worth telling people to have the cli version from this week so it auto installs the plugin.

> Historically, we have offered a separate [tf2pulumi](https://github.com/pulumi/tf2pulumi) tool to convert snippets of Terraform to Pulumi. Today we're announcing a new converter that supports converting the vast majority of complete Terraform projects, including modules. There's no longer a separate tool to download -- you can run the new converter directly from the Pulumi CLI with the `pulumi convert --from terraform` command to convert Terraform to Pulumi TypeScript, Python, Go, or C#.

Reviewer 2: In the beginning of the blog: "Today, we're excited to announce support" then in the second paragraph "Today we're announcing a new converter". I suggest rewriting the paragraph without repeating:

> Historically, we have offered a separate [tf2pulumi](https://github.com/pulumi/tf2pulumi) tool to convert snippets of Terraform code to Pulumi. However, we realized that having a separate CLI for conversion made this feature less accessible for the majority of Pulumi users who don't want to download `tf2pulumi` and keep it up to date. Instead, you can run the new converter directly from the Pulumi CLI with the `pulumi convert --from terraform` command to convert Terraform to Pulumi TypeScript, Python, Go, or C#. 
> 
> This also means that from now on, it will be much easier for us to deliver updates and improvements to the converter which users will get when they update their Pulumi CLI.

Reviewer 3: Actually convert fixes are now decoupled from cli fixes to some degree. We should make this sound like we can do fixes even easier.

Author: Tweaked the text here so it's not repeating.

## Supported Terraform Features

The following major features are supported:

> * Variables, providers, outputs, resources, and data sources

Reviewer 2: I believe we are dropping provider blocks

Reviewer 3: We are! I can have a quick look this evening if we can cover them. Might be easy to do some common cases.

Reviewer 3: We are! I can have a quick look this evening if we can cover them. Might be easy to do some common cases.

Author: Updated to omit that from the bullet.

* Terraform Modules are converted to Pulumi components
* Almost all HCL2 expression syntax

> * Almost all Terraform meta-arguments are supported

Reviewer 3: Ehh this feels dishonest, we're missing most of them currently. I think it's only count and range that we currently handle. Easy to add to the converter, harder to add to PCL and we're missing engine support for triggers.

Author: Removed this bullet

    name: pulumi.output(args.namePrefix).apply(namePrefix => namePrefix.replace("_", "-")),
```

{{% /choosable %}}
{{% choosable language python %}}

```python
alb = aws.lb.LoadBalancer(f"{name}-alb",
    name=pulumi.Output.from_input(args["namePrefix"]).apply(lambda np: np.replace("_", "-")),
```

{{% /choosable %}}
{{% choosable language go %}}

```go
alb, err := lb.NewLoadBalancer(ctx, fmt.Sprintf("%s-alb", name), &lb.LoadBalancerArgs{
    Name: args.NamePrefix.ToStringOutput().ApplyT(func(name string) string {
        return strings.ReplaceAll(name, "_", "-")
    }).(pulumi.StringOutput),
```

{{% /choosable %}}
{{% choosable language csharp %}}

```csharp
var alb = new Aws.Lb.LoadBalancer($"{name}-alb", new()
{
    Name = args.NamePrefix.Apply(name => name.Replace("_", "-"));
```

{{% /choosable %}}
{{< /chooser >}}

There are some other TODOs for missing `merge` and `template_file` support, which also need to be filled in. After addressing these and some other tweaks to make the code compile, we can now run the Pulumi program with `pulumi up` to provision the Vault installation with Pulumi.

JVP: TODO: Show how to fix `merge` and `template_file`.

> The converter has saved us a ton of time, converting over 1,000 lines of Terraform to a modern Pulumi language, with only a small number of manual fix-ups. From here, we can leverage our IDE and compiler to further refactor and improve the code, one of the many benefits of Pulumi!

Reviewer 2: Does it make sense here to talk about what's next in terms of importing the state to Pulumi or do we want a separate blog post on that when it lands in the converter?

Reviewer 3: I'd leave that to when we do the state converter. Maybe a small comment about running import manually?

Author: Luke also was wondering about this as it's likely a common question we'll get. I added a section on it at the end on what can be done today and teasing built-in support coming soon.

---
title: "Converting Full Terraform Programs to Pulumi"
date: 2023-06-12
meta_desc: Learn how to convert whole Terraform programs to Pulumi using the new Terraform converter

meta_image: meta.png

authors:
    - justin-vanpatten

tags:
    - migration
---

> 

Reviewer 4: ```suggestion

Over the last 2 years, we've seen an increasing trend of cloud development teams migrating to Pulumi from Terraform. These teams often have experience with and meaningful investment in Terraform, but have also typically run into limits of expressivity, productivity, scalability or reliability with their existing tools. One of the first questions we hear when they decide to move to Pulumi is "how will I migrate my existing Terraform projects over?".

```

> Today, we're excited to announce support for converting whole Terraform projects to Pulumi via the `pulumi convert` command in the Pulumi CLI. The new Terraform converter includes support for Terraform modules, core features of Terraform 1.4, and the majority of Terraform built-in functions, converting to Pulumi TypeScript, Python, Go, or C#. The new converter can significantly reduce the amount of time it takes to migrate Terraform to Pulumi. Let's dig in to learn more about the new converter and how to use it.

Reviewer 4: ```suggestion
Today, we're excited to announce new support for converting whole Terraform projects to Pulumi via the `pulumi convert` command in the Pulumi CLI. The new Terraform converter includes support for Terraform modules, core features of Terraform 1.4, and the majority of Terraform built-in functions, converting to Pulumi TypeScript, Python, Go, or C#. The new converter can significantly reduce the amount of time it takes to migrate Terraform to Pulumi. Let's dig in to learn more about the new converter and how to use it.
```

<!--more-->

Historically, we have offered a separate [tf2pulumi](https://github.com/pulumi/tf2pulumi) tool to convert small snippets of Terraform to Pulumi. The new converter is no longer a separate tool. As of [v3.71.0](/docs/install/), you can run the new converter directly from the Pulumi CLI with the `pulumi convert --from terraform` command. And you can convert more than small snippets -- the new converter supports converting full Terraform programs.

> 

Reviewer 4: ```suggestion

The new support in `pulumi convert` builds upon Pulumi's [CrossCode](/crosscode/) foundations for providing universal infrastructure as code support across a wide variety of programming languages and converstion tooling between them. It also introduces a new concept of `converter` plugin in the Pulumi engine, which allows conversion tools from other Infrastructure as Code platforms to be integrated into the same `pulumi convert` experience in the future, both as part of the core project, as well as by other ecosystem partners and contributors.

Several common use cases are supported via the new `pulumi convert --from terraform` support in the Pulumi CLI:

* Converting your organization's existing Terraform projects to Pulumi
* Converting your organization's existing Terraform modules to Pulumi, to be consumed as part of existing Pulumi projects
* Converting 3rd party open source Terraform modules or projects which address a use case you want to incorporate into your existing Pulumi projects

```

First, clone the repo and `cd` into the directory containing the Terraform project:

```bash
$ git clone https://github.com/avantoss/vault-infra.git
$ cd vault-infra/terraform/main
```

Next, run the converter:

{{< chooser language "typescript,python,go,csharp" >}}
{{% choosable language typescript %}}

```bash
$ pulumi convert --from terraform --language typescript --out pulumi
```

{{% /choosable %}}
{{% choosable language python %}}

```bash
$ pulumi convert --from terraform --language python --out pulumi
```

{{% /choosable %}}
{{% choosable language go %}}

```bash
$ pulumi convert --from terraform --language go --out pulumi
```

{{% /choosable %}}
{{% choosable language csharp %}}

```bash
$ pulumi convert --from terraform --language csharp --out pulumi
```

{{% /choosable %}}
{{< /chooser >}}

![pulumi convert](pulumi-convert-from-terraform.gif)

> The converted code is generated in the specified `pulumi` output directory. There are two primary files generated from this program's conversion:

Reviewer 4: ```suggestion
The converted code is generated in the specified `pulumi` output directory. A complete Pulumi project is generated, including two primary code files specific to this this program's conversion:
```

* {{< langfile >}} contains the converted code for the main program
* <pulumi-chooser type="language" options="typescript,python,go,csharp" option-style="none" class="inline">
    <pulumi-choosable type="language" value="typescript"><code>vault.ts</code></pulumi-choosable>
    <pulumi-choosable type="language" value="python"><code>vault.py</code></pulumi-choosable>
    <pulumi-choosable type="language" value="go"><code>vault.go</code></pulumi-choosable>
    <pulumi-choosable type="language" value="csharp"><code>vault.cs</code></pulumi-choosable>
  </pulumi-chooser> contains the <code>Vault</code> Pulumi component, converted from the Terraform module

> ### Addressing TODOs

Reviewer 4: I think this is okay as is - but one thought is that we go into a lot of detail here - and possibly could move some of this into reference docs.  I could imagine having just one of these examples here, and then keeping the transition into "After addressing these and some other tweaks so that the code compiles."

This is especially true, as I imagine it will be fairly soon that we remove the need for there?

But possibly easy enough to just leave as is, and remove them over the coming weeks as we address the underlying gaps?

Reviewer 3: > This is especially true, as I imagine it will be fairly soon that we remove the need for there?

Yes! I'm pretty sure fixing "replace" is just a case of getting a new std release and telling the converter to use it, so will probably be in next week (Maybe even for the main release on Monday)

Author: Moved to just a single example for the blog post. I'll save the rest for a more detailed walkthrough in reference docs.

The file for the `Vault` component makes up the bulk of the implementation and contains some TODOs emitted by the converter that must be addressed manually.

#### Replace

For example, the converter doesn't yet support the `replace` built-in function.

{{< chooser language "typescript,python,go,csharp" >}}
{{% choosable language typescript %}}

```typescript
name: notImplemented("replace(var.name_prefix,\"_\",\"-\")"),
```

{{% /choosable %}}
{{% choosable language python %}}

```python
name=not_implemented("replace(var.name_prefix,\"_\",\"-\")"),
```

{{% /choosable %}}
{{% choosable language go %}}

```go
Name: notImplemented("replace(var.name_prefix,\"_\",\"-\")"),
```

{{% /choosable %}}
{{% choosable language csharp %}}

```csharp
Name = NotImplemented("replace(var.name_prefix,\"_\",\"-\")"),
```

{{% /choosable %}}
{{< /chooser >}}

> We can fill in an implementation.

Reviewer 4: ```suggestion
We can fill in an implementation.  Note that we get to use the full expressiveness of the native and familiar string manipulation libraries in our target programming language, instead of the relatively constrained options of the Terraform DSL.
```

```

{{% /choosable %}}
{{% choosable language csharp %}}

```csharp
var userdata = Output.Tuple(args.NamePrefix, args.Region, (Input<string>)vaultResources.Id, args.VaultConfigDir).Apply(t =>
{
   var (namePrefix, region, vaultResourcesId, vaultConfigDir) = t;
   return $@"#!/bin/bash

# Get the Instance ID
INSTANCE_ID=$(curl http://169.254.169.254/latest/meta-data/instance-id)

# Set the Hostname
hostnamectl set-hostname ""{ namePrefix }-$INSTANCE_ID""
systemctl restart rsyslog.service

# Get Configuration and SSL Certs
aws --region { region } s3 cp s3://{ vaultResourcesId }/resources/config/config.hcl { vaultConfigDir }/config.hcl

...
";
});
```

{{% /choosable %}}
{{< /chooser >}}

### Wrapping Up

After addressing these and some other tweaks so that the code compiles, we can now run the converted program with `pulumi up` to provision the Vault installation with Pulumi.

The converter has saved us a ton of time, converting over 1,000 lines of Terraform to a modern Pulumi language, with only a small number of manual fix-ups required. From here, we can leverage our IDE and compiler to further refactor and improve the code, one of the many benefits of Pulumi!

> 

Reviewer 4: Is there anything we can/should say about migrating state over?  I think we ended up with not really having a great story here yet.  What's the best we could suggest?  And if nothing, should we reference that this is coming soon?

I suspect in practice, this will be one of the top questions we get.

Reviewer 3: I think we should say we're working on it and it's coming very soon. Should probably be the next target feature to release, I don't think it will be a lot of work a lot of the groundwork is in place already.

Author: I added a section on this, mentioning what you can do today with some existing helper code we have for TypeScript and Go (linking to the reference docs on that; though, need to update those docs to mention Go), and teasing that built-in support is coming soon.

## Get Started

> Support for the new `pulumi convert --from terraform` command is available today in v3.71.0 of the Pulumi CLI. We are actively addressing issues, adding support for missing features, and improving the quality of the converted code, to further reduce the amount of manual fix-ups required when doing a migration.

Reviewer 4: ```suggestion
Support for the new `pulumi convert --from terraform` command is available today in v3.71.0 of the Pulumi CLI. [Download](/docs/install/) the latest Pulumi CLI and give the new converter a try today. If you run into any issues, please [let us know](https://github.com/pulumi/pulumi/issues/new/choose) or reach out in the [Pulumi Community Slack](https://slack.pulumi.com) with any questions!
```

> [Download](/docs/install/) the latest Pulumi CLI and give the new converter a try today. If you run into any issues, please [let us know](https://github.com/pulumi/pulumi/issues/new/choose)!

Reviewer 4: ```suggestion
```
